### PR TITLE
Split into polyfill-library and polyfill-service

### PR DESCRIPTION
**Pull request by marionebl**
Promised in #1267, this is a first try at splitting into two distinct packages while 

* avoiding lerna
* keeping a single repo

The intended improvements are:

* smaller install size
* easier dependency management

I tested as much of the build system as I could but you certainly should take all of the scripts for another spin as I do not have a Browser Stack, Pingdom or Fastly account.

--
**Side note**: I'm aware this is very hard to review. Presumably this is a very first start of the discussion and the core team will implement something like this itself.

### CODE REVIEW

**Comment by JakeChampion** [comment]
@marionebl I will think over the implications of this PR this weekend. I really appreciate the work you have put in to this in such a short amount of time.

If other people are reviewing this, you might find it simpler to review the change via the [raw patch file](https://patch-diff.githubusercontent.com/raw/Financial-Times/polyfill-service/pull/1267.patch) instead of using Github's diff mode.

File: package.json [hunk 1]
```diff
@@ -1,8 +1,16 @@
 {
   "name": "polyfill-service",
+  "version": "0.0.0",
+  "private": true,
   "description": "A polyfill combinator",
-  "main": "lib/index.js",
-  "bin": "bin/polyfill-service",
```
Comment (JakeChampion) [inline package.json:5]: Removing the bin, main and npm start script would all count as a breaking change.

File: packages/polyfill-library/polyfills/Array/of/polyfill.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/*! https://mths.be/array-of v0.1.0 by @mathias */
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/Array/of/polyfill.js]: This file can be deleted as it is automatically built from the `array.of` npm package. I'm not too sure how it ended up being commited as it is included in the [.gitignore](https://github.com/Financial-Times/polyfill-service/pull/1267/files#diff-a084b794bc0759e7a6b77810e01874f2R11) file.

File: packages/polyfill-library/polyfills/AudioContext/polyfill.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
+(function () {
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/AudioContext/polyfill.js]: Same with this file, safe to delete as it is built during from an npm package when running `npm install`. This file is also included in the .gitignore file, perhaps this was commited before the .gitignore file was updated?

File: packages/polyfill-library/polyfills/EventSource/polyfill.js [hunk 4]
```diff
@@ -0,0 +1,683 @@
+/** @license
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/EventSource/polyfill.js]: Safe to delete this file, it is built during `npm install`.

File: packages/polyfill-library/polyfills/HTMLPictureElement/polyfill.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+/*! picturefill - v3.0.2 - 2016-02-12
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/HTMLPictureElement/polyfill.js]: Safe to delete this file, it is built on demand when running `npm install`

File: packages/polyfill-library/polyfills/IntersectionObserver/polyfill.js [hunk 6]
```diff
@@ -0,0 +1,661 @@
+/**
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/IntersectionObserver/polyfill.js]: Safe to delete this file, it is built via `npm install`

File: packages/polyfill-library/polyfills/Intl/polyfill.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
...["ayl","ar"],ayn:["ayn","ar"],ayp:["ayp","ar"],bbz:["bbz","ar"],bfi:["bfi","sgn"],bfk:["bfk","sgn"],bjn:["bjn","ms"],bog:["bog","sgn"],bqn:["bqn","sgn"],bqy:["bqy","sgn"],btj:["btj","ms"],bve:["bve","ms"],bvl:["bvl","sgn"],bvu:["bvu","ms"],bzs:["bzs","sgn"],cdo:["cdo","zh"],cds:["cds","sgn"],cjy:["cjy","zh"],cmn:["cmn","zh"],coa:["coa","ms"],cpx:["cpx","zh"],csc:["csc","sgn"],csd:["csd","sgn"],cse:["cse","sgn"],csf:["csf","sgn"],csg:["csg","sgn"],csl:["csl","sgn"],csn:["csn","sgn"],csq:["csq","sgn"],csr:["csr","sgn"],czh:["czh","zh"],czo:["czo","zh"],doq:["doq","sgn"],dse:["dse","sgn"],dsl:["dsl","sgn"],dup:["dup","ms"],ecs:["ecs","sgn"],esl:["esl","sgn"],esn:["esn","sgn"],eso:["eso","sgn"],eth:["eth","sgn"],fcs:["fcs","sgn"],fse:["fse","sgn"],fsl:["fsl","sgn"],fss:["fss","sgn"],gan:["gan","zh"],gds:["gds","sgn"],gom:["gom","kok"],gse:["gse","sgn"],gsg:["gsg","sgn"],gsm:["gsm","sgn"],gss:["gss","sgn"],gus:["gus","sgn"],hab:["hab","sgn"],haf:["haf","sgn"],hak:["hak","zh"],hds:["hds","sgn"],hji:["hji","ms"],hks:["hks","sgn"],hos:["hos","sgn"],hps:["hps","sgn"],hsh:["hsh","sgn"],hsl:["hsl","sgn"],hsn:["hsn","zh"],icl:["icl","sgn"],ils:["ils","sgn"],inl:["inl","sgn"],ins:["ins","sgn"],
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/Intl/polyfill.js]: Safe to delete this file, it is built via `npm install`

File: packages/polyfill-library/polyfills/JSON/polyfill.js [hunk 8]
```diff
@@ -0,0 +1,902 @@
+/*! JSON v3.3.2 | http://bestiejs.github.io/json3 | Copyright 2012-2014, Kit Cambridge | http://kit.mit-license.org */
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/JSON/polyfill.js]: Safe to delete this file, it is built via `npm i`

File: packages/polyfill-library/polyfills/MutationObserver/polyfill.js [hunk 9]
```diff
@@ -0,0 +1,620 @@
+/*!
```
Comment (JakeChampion) [inline packages/polyfill-library/polyfills/MutationObserver/polyfill.js]: Safe to delete this file.

File: packages/polyfill-library/package.json [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    "handlebars": "^4.0.6",
+    "lazystream": "^1.0.0",
+    "lodash": "^4.17.4",
+    "lru-cache": "^4.0.2",
+    "merge2": "^1.0.3",
+    "semver": "^5.3.0",
+    "stream-from-promise": "^1.0.0",
+    "stream-to-string": "^1.1.0",
+    "tsort": "0.0.1",
+    "useragent": "^2.1.12"
```
Comment (JakeChampion) [inline packages/polyfill-library/package.json]: missing usertiming npm package

**Comment by JakeChampion** [comment]
@marionebl I removed all the polyfills which are automatically built during `npm install`, this brought the diff down by quite a bit :-)

File: package.json (hunk 1 above)
Comment (marionebl) [reply package.json:5]: I'd move them to the sub manifests:

`main` ist already in [polyfill-library](https://github.com/Financial-Times/polyfill-service/pull/1267/files/f426ca1e594e0c29e1bf55be460ad5bf0d3c2b57#diff-d3a4d0e3a2b26bc7bdf9e9a9aca2e3d3).

`bin` should be added to [polyfill-service](https://github.com/Financial-Times/polyfill-service/pull/1267/files/f426ca1e594e0c29e1bf55be460ad5bf0d3c2b57#diff-d3f07169ee33bf8a4b404862a89da0b3)

**Comment by marionebl** [comment]
Hey @JakeChampion, can I help this along in some way?

**Comment by JakeChampion** [comment]
@marionebl Hi, sorry this is taking a while to get through, we have some issues with our testing environment and some polyfills which are failing on master. I'm looking into this issues, when they are resolved I will look at getting this PR back on track.

**Comment by marionebl** [comment]
@JakeChampion I'd be happy to help this along with additional work. Anything you need some help with?

**Comment by marionebl** [comment]
Sorry to nag - scanning through my long-living PRs and found this. I'd love to land this at some point. Is there anything I can do to help?

**Comment by marionebl** [comment]
Hey there, I'll make an effort to port the changes that happened in the meantime to `@marionebl/polyfill-service`, which would enable a new merge conflict free PR. Would that increase the chance to get this merged?

**Comment by JakeChampion** [comment]
@marionebl Yes, we are happy to merge this work once it is ready

**Comment by marionebl** [comment]
@JakeChampion I resolved the merge conflicts that cropped up in the meantime as best as I could.

I guess there are some todos I can't solve due to the circleci being intransparet to me. (Any chance for me to see the build logs?)

**Comment by marionebl** [comment]
@JakeChampion Any chance we can land this in say the next 3 months? I'd go with a different approach to extract the polyfills if not.

**Comment by marionebl** [comment]
Closing this as I don't see a real chance for this type of change to land. Will create a distinct project to extract and publish the `polyfill-library` from this one.

**Comment by SebastianS90** [comment]
As author of [`webpack-polyfill-injector`](https://github.com/SebastianS90/webpack-polyfill-injector) I was hoping for this to get merged :cry:

@marionebl can you please comment here when you created a separate project for `polyfill-library`, or tag me somewhere in the new project? I can also offer some help to maintain the library project.

If have two ideas in mind how the library project could be done without too much manual work:

**Idea 1: Extracting polyfills into separate project**
The new project ships the polyfill sources (i.e. the `polyfills` folder here). A daily cronjob could automatically check for new versions of `polyfill-service`, copy over the polyfills, run the build script and publish a new version if it succeeds. If the build fails, a team member is notified to manually update the library project (probably there was a dependency added in `polyfill-service` that should also be added to `polyfill-library`)

**Idea 2: No polyfills in the library project itself, only published in binary form**
This idea somehow cheats npm. Add `polyfill-service` as devDependency locked to the current version. The build script for the library just copies the `__dist` folder from `polyfill-service` to the current working dir and the files there are uploaded when doing `npm publish`. When someone installs `polyfill-library`, then the already compiled `__dist` folder is downloaded from npm, but neither `polyfill-service` nor all its dependencies are installed because `polyfill-service` is only a devDependency of `polyfill-library`.
The update process here would look as follows: Whenever there is a new version of `polyfill-service`, we need to change `package.json` in `polyfill-library` to increment its own version number and update the devDependency for `polyfill-service` to the new version. No problems are expected here, except when `polyfill-service` decides to rename the `__dist` folder.


In case you decide to host the `polyfill-library` on GitLab.com, then you can use their [pipeline schedules](https://docs.gitlab.com/ce/user/project/pipelines/schedules.html) to automatically run a daily tasks which updates the project.

As I said, I'd be happy to help you there, I might also be able to do it alone. But I don't want to steal your general idea or the `polyfill-library` project name. Just let me know what you think.

**Comment by JakeChampion** [comment]
@marionebl @SebastianS90, I've made an updated version of the pull request (https://github.com/Financial-Times/polyfill-service/pull/1693). Apologies for letting this one take so long to get through.

**Pull request closed without merging**